File: backend/app/api/regulatory.py
"""Regulatory parameter endpoints (SRS 1.4, ADR-007).

An administration surface, not a business one: what limits are loaded, since when, from which
norm, and — the field that matters most — which of them nobody has verified against the
official text yet. That list is what a deployment checklist has to be empty of.
"""

from __future__ import annotations

from datetime import date
from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.auth.dependencies import current_principal, require_roles
from app.auth.principal import Role
from app.infra.database import get_session
from app.regulatory.loader import unverified_codes
from app.regulatory.rules import REQUIRED_PARAMETER_CODES
from app.regulatory.service import (
    OverlappingPeriodError,
    UnknownParameterError,
    UnverifiedParameterError,
    history,
    known_codes,
    parameter_in_force,
    revisions,
    set_parameter,
    verify_parameter,
)

# Los parámetros regulatorios son nacionales, no de una unidad de negocio, así que aquí no hay
# ámbito por unidad — pero sí hace falta estar autenticado, y escribir un límite es un acto de
# administración funcional: quien lo carga afirma haber leído el texto oficial.
router = APIRouter(
    prefix="/api/v1/regulatory", tags=["regulatory"], dependencies=[Depends(current_principal)]
)

SessionDep = Annotated[Session, Depends(get_session)]


def _as_dict(row: Any) -> dict[str, Any]:
    return {
        "code": row.code,
        "value": row.value,
        "unit": row.unit,
        "description": row.description,
        "norm_ref": row.norm_ref,
        "article_ref": row.article_ref,
        "source_url": row.source_url,
        "effective_from": row.effective_from.isoformat(),
        "effective_to": row.effective_to.isoformat() if row.effective_to else None,
        "verified": row.is_verified,
        "verified_by": row.verified_by,
        "verified_at": row.verified_at.isoformat() if row.verified_at else None,
        "strict": row.strict,
        # RF-150's «usuario que modificó». Kept apart from `verified_by`: loading a figure and
        # certifying it against the official text are different claims.
        "created_by": row.created_by,
        "updated_by": row.updated_by,
        "updated_at": row.updated_at.isoformat() if row.updated_at else None,
    }


@router.get("/parameters")
def list_parameters(session: SessionDep) -> dict[str, Any]:
    """What is loaded, what the rules need, and what nobody has verified."""
    loaded = known_codes(session)
    missing = [
        code
        for code in REQUIRED_PARAMETER_CODES
        if not any(existing == code or existing.startswith(f"{code}.") for existing in loaded)
    ]
    return {
        "loaded": loaded,
        "required_by_rules": list(REQUIRED_PARAMETER_CODES),
        # A rule with no parameter reports that it cannot judge; it never passes silently.
        # Surfacing the gap here is how it gets closed before the pilot.
        "missing": missing,
        "unverified": unverified_codes(session),
    }


@router.get("/parameters/{code}")
def parameter_history(session: SessionDep, code: str) -> list[dict[str, Any]]:
    """Every period recorded for a code — what makes an old approval explicable."""
    rows = history(session, code)
    if not rows:
        raise HTTPException(status.HTTP_404_NOT_FOUND, f"no hay parámetros con el código '{code}'")
    return [_as_dict(row) for row in rows]


@router.get("/parameters/{code}/in-force")
def parameter_on_date(
    session: SessionDep,
    code: str,
    on: Annotated[date | None, Query()] = None,
) -> dict[str, Any]:
    """The value that governed a code on a date."""
    try:
        return _as_dict(parameter_in_force(session, code, on=on))
    except UnknownParameterError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    except UnverifiedParameterError as exc:
        # 409, not 404: the value exists and is deliberately not usable yet.
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc)) from exc


class ParameterIn(BaseModel):
    code: str = Field(min_length=1, max_length=128)
    value: Any
    norm_ref: str = Field(min_length=1, max_length=255)
    effective_from: date
    unit: str | None = None
    description: str | None = None
    article_ref: str | None = None
    source_url: str | None = None
    effective_to: date | None = None
    strict: bool = False
    #: Why this row was written or corrected, for whoever reads the revision in a year.
    note: str | None = Field(default=None, max_length=2000)
    # No `verified_by`: a verification the caller could type is a verification nobody made. It is
    # its own endpoint below, and the person who read the official text is whoever the token says
    # they are. The seed loader still passes a name because it runs at a console with no token —
    # and it only certifies the entries the file itself marks verified.


@router.post(
    "/parameters",
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_roles(Role.FUNCTIONAL_ADMIN, Role.IT_ADMIN))],
)
def upsert_parameter(
    session: SessionDep,
    payload: ParameterIn,
    principal: Annotated[Any, Depends(current_principal)] = None,
) -> dict[str, Any]:
    """Record a regulatory value, closing the previous period rather than replacing it."""
    try:
        row = set_parameter(
            session,
            code=payload.code,
            value=payload.value,
            norm_ref=payload.norm_ref,
            effective_from=payload.effective_from,
            unit=payload.unit,
            description=payload.description,
            article_ref=payload.article_ref,
            source_url=payload.source_url,
            effective_to=payload.effective_to,
            strict=payload.strict,
            actor=principal.subject,
            note=payload.note,
        )
    except OverlappingPeriodError as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc)) from exc
    session.commit()
    return _as_dict(row)


@router.get(
    "/parameters/{code}/revisions", summary="Quién modificó qué, y desde qué valor (RF-150)"
)
def parameter_revisions(session: SessionDep, code: str) -> list[dict[str, Any]]:
    """The history of the *edits*, next to the history of the values the periods already hold.

    Oldest first: the answer is a story, and a story read backwards is a different story.
    """
    return [
        {
            "at": row.at.isoformat(),
            "action": row.action,
            "actor": row.actor,
            "changed": row.changed,
            "note": row.note,
        }
        for row in revisions(session, code)
    ]


class VerifyIn(BaseModel):
    """Which period is being certified. No `verified_by`: that is the token's subject."""

    effective_from: date
    note: str | None = Field(default=None, max_length=2000)


@router.post(
    "/parameters/{code}/verify",
    dependencies=[Depends(require_roles(Role.FUNCTIONAL_ADMIN, Role.IT_ADMIN))],
    summary="Declarar que se leyó el texto oficial y el valor coincide (ADR-007)",
)
def verify(
    session: SessionDep,
    code: str,
    payload: VerifyIn,
    principal: Annotated[Any, Depends(current_principal)] = None,
) -> dict[str, Any]:
    """Its own endpoint, not a field on the write: loading a figure and certifying it are two acts,
    and folding them together is how a bulk import marks seven values verified for one flag."""
    try:
        row = verify_parameter(
            session,
            code,
            effective_from=payload.effective_from,
            actor=principal.subject,
            note=payload.note,
        )
    except UnknownParameterError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    session.commit()
    return _as_dict(row)

File: web/src/api/regulatory.ts
/**
 * Regulatory parameter client (RF-150, ADR-007).
 *
 * The endpoints are not scoped to a business unit: the limits a regulator publishes are national
 * (ADR-009 keeps nothing corporate inside a unit's data). So nothing here takes a unit, which is
 * itself the thing worth noticing when reading this file next to the others.
 */

import { ApiError } from './planning';
import { authHeaders, notifyExpired } from './session';

export interface Parameter {
  code: string;
  /** Stored as an object so JSONB querying is uniform; `{v: 24}` for a scalar. */
  value: Record<string, unknown>;
  unit: string | null;
  description: string | null;
  norm_ref: string;
  article_ref: string | null;
  source_url: string | null;
  effective_from: string;
  effective_to: string | null;
  /** True only when a person stated they read the official text (ADR-007). */
  verified: boolean;
  verified_by: string | null;
  verified_at: string | null;
  strict: boolean;
  created_by: string | null;
  updated_by: string | null;
  updated_at: string | null;
}

export interface ParameterIndex {
  loaded: string[];
  required_by_rules: string[];
  /** Codes a rule needs and nobody loaded. The list a deployment checklist has to empty. */
  missing: string[];
  /** In force but unverified against the official text. The other list to empty. */
  unverified: string[];
}

export interface Revision {
  at: string;
  action: string;
  actor: string;
  /** {field: {from, to}}. Empty for a creation, where everything is new. */
  changed: Record<string, { from: unknown; to: unknown }>;
  note: string | null;
}

const BASE = '/api/v1/regulatory';

async function call<T>(path: string, init: RequestInit = {}): Promise<T> {
  const response = await fetch(path, {
    ...init,
    headers: {
      ...authHeaders(),
      ...(init.body ? { 'Content-Type': 'application/json' } : {}),
      ...(init.headers ?? {}),
    },
  });
  if (!response.ok) {
    if (response.status === 401) notifyExpired();
    let detail = `${response.status} ${response.statusText}`;
    try {
      const body = (await response.json()) as { detail?: unknown };
      if (typeof body.detail === 'string') detail = body.detail;
    } catch {
      // No JSON body; the status line stands.
    }
    throw new ApiError(response.status, detail);
  }
  return (await response.json()) as T;
}

export function fetchIndex(signal?: AbortSignal): Promise<ParameterIndex> {
  return call<ParameterIndex>(`${BASE}/parameters`, { signal });
}

export function fetchHistory(code: string, signal?: AbortSignal): Promise<Parameter[]> {
  return call<Parameter[]>(`${BASE}/parameters/${encodeURIComponent(code)}`, { signal });
}

export function fetchRevisions(code: string, signal?: AbortSignal): Promise<Revision[]> {
  return call<Revision[]>(`${BASE}/parameters/${encodeURIComponent(code)}/revisions`, { signal });
}

/**
 * Record that a person read the official text and the value matches.
 *
 * No author parameter: the server signs it with the token's subject. A verification the browser
 * could fill in would be a verification nobody made.
 */
export function verifyParameter(
  code: string,
  effectiveFrom: string,
  note?: string,
): Promise<Parameter> {
  return call<Parameter>(`${BASE}/parameters/${encodeURIComponent(code)}/verify`, {
    method: 'POST',
    body: JSON.stringify({ effective_from: effectiveFrom, note: note ?? null }),
  });
}
